contrib/generate-api-changes.sh

```diff
@@ -0,0 +1,144 @@
+#!/usr/bin/env bash
+#
+# Creates a temporary branch with an api-changes patch for each commit back to master.
+
+set -euo pipefail
+
+REPO_DIR=$(git rev-parse --show-toplevel)
+API_DIR="$REPO_DIR/api"
+
+NIGHTLY=$(cat nightly-version)
+# Our docs have broken intra doc links if all features are not enabled.
+RUSTDOCFLAGS="-A rustdoc::broken_intra_doc_links"
+
+# `sort -n -u` doesn't work for some reason.
+SORT="sort --numeric-sort"
+
+# Get temporary branch name.
+TEMP_BRANCH=$(mktemp -u temp-branch-XXXXXX)
+
+# Sort order is effected by locale. See `man sort`.
+# > Set LC_ALL=C to get the traditional sort order that uses native byte values.
+export LC_ALL=C
+
+main() {
+    need_nightly
+
+    # Create a new temporary branch
+    current_branch=$(git branch --show-current)
+
+    # Find commits on current branch that are not in master
+    commits=$(git rev-list master.."$current_branch")
+
+    git checkout -b "$TEMP_BRANCH" master || exit 1
+
+    mkdir "api"
+    for crate in base58 bitcoin hashes io units; do
+        mkdir "api/$crate"
+    done
+
+    # Reverse the commit list to process from oldest to newest
+    commits=$(echo "$commits" | tac)
+
+    # Iterate through each commit, checking API before cherry picking the commit.
+    for commit in $commits; do
+        create_api_patch
+        git cherry-pick "$commit"
+    done
+
+    # And do the final check.
+    create_api_patch
+}
+
+create_api_patch() {
+    check_api
+    commit_changes
+}
+
+check_api() {
+    generate_api_files_base58
+    generate_api_files_bitcoin
+
+    # These ones have an "alloc" feature we want to check.
+    generate_api_files "hashes"
+    generate_api_files "io"
+    generate_api_files "units"
+}
+
+commit_changes() {
+    pushd "$REPO_DIR" > /dev/null
+
+    if [[ $(git status --porcelain api) ]]; then
+        git add -A
+        git commit -m "api: Run check-for-api-changes.sh" -n
```

Comment:
In f84fb80c6d894b49c2f1d909bf6ad865a908bcd2:

I'd also change the author field and add `--no-gpg-sign`.